[SPARK-12594] [SQL] Outer Join Elimination by Filter Conditions

gatorsmile opened the pull request:
Conversion of outer joins, if the predicates in filter conditions can restrict the result sets so that all null-supplying rows are eliminated. 
- `full outer` -> `inner` if both sides have such predicates
- `left outer` -> `inner` if the right side has such predicates
- `right outer` -> `inner` if the left side has such predicates
- `full outer` -> `left outer` if only the left side has such predicates
- `full outer` -> `right outer` if only the right side has such predicates

If applicable, this can greatly improve the performance, since outer join is much slower than inner join, full outer join is much slower than left/right outer join.

gatorsmile committed:
Merge remote-tracking branch 'upstream/master' into outerJoinElimination

gatorsmile committed:
outer join conversion

gatorsmile committed:
outer join conversion

SparkQA commented:
**[Test build #48559 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/48559/consoleFull)** for PR 10542 at commit [`90576aa`](https://github.com/apache/spark/commit/90576aacb085e9192360505ac5687a3df5e54b6f).

JoshRosen committed:
[SPARK-10359][PROJECT-INFRA] Multiple fixes to dev/test-dependencies.sh script

This patch includes multiple fixes for the `dev/test-dependencies.sh` script (which was introduced in #10461):

- Use `build/mvn --force` instead of `mvn` in one additional place.
- Explicitly set a zero exit code on success.
- Set `LC_ALL=C` to make `sort` results agree across machines (see https://stackoverflow.com/questions/28881/).
- Set `should_run_build_tests=True` for `build` module (this somehow got lost).

Author: Josh Rosen <joshrosen@databricks.com>

Closes #10543 from JoshRosen/dep-script-fixes.

gatorsmile committed:
added test cases.

gatorsmile committed:
added test cases.

SparkQA commented:
**[Test build #48559 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/48559/consoleFull)** for PR 10542 at commit [`90576aa`](https://github.com/apache/spark/commit/90576aacb085e9192360505ac5687a3df5e54b6f).
 * This patch **fails from timeout after a configured wait of \`250m\`**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/48559/
Test FAILed.

SparkQA commented:
**[Test build #48564 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/48564/consoleFull)** for PR 10542 at commit [`192ab19`](https://github.com/apache/spark/commit/192ab19337e0d9dc93eacf69b89ef5d105fe36ae).

SparkQA commented:
**[Test build #48564 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/48564/consoleFull)** for PR 10542 at commit [`192ab19`](https://github.com/apache/spark/commit/192ab19337e0d9dc93eacf69b89ef5d105fe36ae).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/48564/
Test PASSed.

viirya committed:
[SPARK-11743][SQL] Move the test for arrayOfUDT

A following pr for #9712. Move the test for arrayOfUDT.

Author: Liang-Chi Hsieh <viirya@gmail.com>

Closes #10538 from viirya/move-udt-test.

Commit:
[SPARK-3873][MLLIB] Import order fixes.

A slight adjustment to the checker configuration was needed; there is
a handful of warnings still left, but those are because of a bug in
the checker that I'll fix separately (before enabling errors for the
checker, of course).

Author: Marcelo Vanzin <vanzin@cloudera.com>

Closes #10535 from vanzin/SPARK-3873-mllib.

viirya committed:
[SPARK-12409][SPARK-12387][SPARK-12391][SQL] Refactor filter pushdown for JDBCRDD and add few filters

This patch refactors the filter pushdown for JDBCRDD and also adds few filters.

Added filters are basically from #10468 with some refactoring. Test cases are from #10468.

Author: Liang-Chi Hsieh <viirya@gmail.com>

Closes #10470 from viirya/refactor-jdbc-filter.

rxin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -769,6 +770,63 @@ object ReorderJoin extends Rule[LogicalPlan] with PredicateHelper {
+ *
+ * This rule should be executed before pushing down the Filter
+ */
+object OuterJoinConversion extends Rule[LogicalPlan] with PredicateHelper {
```
is there a better name for this? this is a form of strength reduction right?

gatorsmile replied:
You are right. `OuterJoinElimination` might sound better. Let me rename it today. 

Since `full outer` is `union distinct` of `left outer` and `right outer`, we are removing `right outer` from `full outer` in conversion from `full outer` to `left outer`.

gatorsmile committed:
renaming

gatorsmile committed:
renaming

SparkQA commented:
**[Test build #48569 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/48569/consoleFull)** for PR 10542 at commit [`c04b53b`](https://github.com/apache/spark/commit/c04b53b0da168bba2a50010d42a13c04efc9a2bd).

liancheng committed:
[SPARK-12592][SQL][TEST] Don't mute Spark loggers in TestHive.reset()

There's a hack done in `TestHive.reset()`, which intended to mute noisy Hive loggers. However, Spark testing loggers are also muted.

Author: Cheng Lian <lian@databricks.com>

Closes #10540 from liancheng/spark-12592.dont-mute-spark-loggers.

rxin committed:
Disable test-dependencies.sh.

Commit:
[SPARK-12286][SPARK-12290][SPARK-12294][SPARK-12284][SQL] always output UnsafeRow

It's confusing that some operator output UnsafeRow but some not, easy to make mistake.

This PR change to only output UnsafeRow for all the operators (SparkPlan), removed the rule to insert Unsafe/Safe conversions. For those that can't output UnsafeRow directly, added UnsafeProjection into them.

Closes #10330

cc JoshRosen rxin

Author: Davies Liu <davies@databricks.com>

Closes #10511 from davies/unsafe_row.

SparkQA commented:
**[Test build #48569 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/48569/consoleFull)** for PR 10542 at commit [`c04b53b`](https://github.com/apache/spark/commit/c04b53b0da168bba2a50010d42a13c04efc9a2bd).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/48569/
Test PASSed.

rxin committed:
Revert "[SPARK-12286][SPARK-12290][SPARK-12294][SPARK-12284][SQL] always output UnsafeRow"

This reverts commit 0da7bd50ddf0fb9e0e8aeadb9c7fb3edf6f0ee6e.

hvanhovell committed:
[SPARK-12362][SQL][WIP] Inline Hive Parser

This PR inlines the Hive SQL parser in Spark SQL.

The previous (merged) incarnation of this PR passed all tests, but had and still has problems with the build. These problems are caused by a the fact that - for some reason - in some cases the ANTLR generated code is not included in the compilation fase.

This PR is a WIP and should not be merged until we have sorted out the build issues.

Author: Herman van Hovell <hvanhovell@questtec.nl>
Author: Nong Li <nong@databricks.com>
Author: Nong Li <nongli@gmail.com>

Closes #10525 from hvanhovell/SPARK-12362.

HyukjinKwon committed:
[SPARK-10180][SQL] JDBC datasource are not processing EqualNullSafe filter

This PR is followed by https://github.com/apache/spark/pull/8391.
Previous PR fixes JDBCRDD to support null-safe equality comparison for JDBC datasource. This PR fixes the problem that it can actually return null as a result of the comparison resulting error as using the value of that comparison.

Author: hyukjinkwon <gurwls223@gmail.com>
Author: HyukjinKwon <gurwls223@gmail.com>

Closes #8743 from HyukjinKwon/SPARK-10180.

srowen committed:
[SPARK-12481][CORE][STREAMING][SQL] Remove usage of Hadoop deprecated APIs and reflection that supported 1.x

Remove use of deprecated Hadoop APIs now that 2.2+ is required

Author: Sean Owen <sowen@cloudera.com>

Closes #10446 from srowen/SPARK-12481.

gatorsmile committed:
extend the condition to cover more cases in non null predicates.

gatorsmile committed:
extend the condition to cover more cases in non null predicates.

SparkQA commented:
**[Test build #48597 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/48597/consoleFull)** for PR 10542 at commit [`65f9125`](https://github.com/apache/spark/commit/65f9125c321d2509dff2663ae9f2db1cb86fb77f).

rxin committed:
[SPARK-12599][MLLIB][SQL] Remove the use of callUDF in MLlib

callUDF has been deprecated. However, we do not have an alternative for users to specify the output data type without type tags. This pull request introduced a new API for that, and replaces the invocation of the deprecated callUDF with that.

Author: Reynold Xin <rxin@databricks.com>

Closes #10547 from rxin/SPARK-12599.

rxin committed:
Revert "Revert "[SPARK-12286][SPARK-12290][SPARK-12294][SPARK-12284][SQL] always output UnsafeRow""

This reverts commit 44ee920fd49d35b421ae562ea99bcc8f2b98ced6.

gatorsmile committed:
added three more expressions: and, or and not

gatorsmile committed:
added three more expressions: and, or and not

gatorsmile committed:
style fix.

gatorsmile committed:
style fix.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/48597/
Test FAILed.

SparkQA commented:
**[Test build #48600 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/48600/consoleFull)** for PR 10542 at commit [`0bb07cb`](https://github.com/apache/spark/commit/0bb07cb7feadf298cc94a3ce044872624177a376).

SparkQA commented:
**[Test build #48600 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/48600/consoleFull)** for PR 10542 at commit [`0bb07cb`](https://github.com/apache/spark/commit/0bb07cb7feadf298cc94a3ce044872624177a376).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/48600/
Test FAILed.

gatorsmile committed:
support non-local predicates and bug fix.

gatorsmile committed:
support non-local predicates and bug fix.

SparkQA commented:
**[Test build #48604 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/48604/consoleFull)** for PR 10542 at commit [`c5ff632`](https://github.com/apache/spark/commit/c5ff632e63b46e9578fed39ce45f79382a87de45).

SparkQA commented:
**[Test build #48604 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/48604/consoleFull)** for PR 10542 at commit [`c5ff632`](https://github.com/apache/spark/commit/c5ff632e63b46e9578fed39ce45f79382a87de45).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/48604/
Test FAILed.

felixcheung committed:
[SPARK-12327][SPARKR] fix code for lintr warning for commented code

shivaram

Author: felixcheung <felixcheung_m@hotmail.com>

Closes #10408 from felixcheung/rcodecomment.

gatorsmile committed:
scala style fix.

gatorsmile committed:
scala style fix.

SparkQA commented:
**[Test build #48609 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/48609/consoleFull)** for PR 10542 at commit [`ee29dd2`](https://github.com/apache/spark/commit/ee29dd230702413238abed77fc6a7691bd4ff6a5).

SparkQA commented:
**[Test build #48609 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/48609/consoleFull)** for PR 10542 at commit [`ee29dd2`](https://github.com/apache/spark/commit/ee29dd230702413238abed77fc6a7691bd4ff6a5).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/48609/
Test FAILed.

thomastechs committed:
[SPARK-12533][SQL] hiveContext.table() throws the wrong exception

Avoiding the the No such table exception and throwing analysis exception as per the bug: SPARK-12533

Author: thomastechs <thomas.sebastian@tcs.com>

Closes #10529 from thomastechs/topic-branch.

gatorsmile commented:
The failed bucket is not caused by the code changes in this PR. Let me submit a PR to fix this Python bug.

gatorsmile commented:
retest it please.

SparkQA commented:
**[Test build #48613 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/48613/consoleFull)** for PR 10542 at commit [`ee29dd2`](https://github.com/apache/spark/commit/ee29dd230702413238abed77fc6a7691bd4ff6a5).

SparkQA commented:
**[Test build #48613 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/48613/consoleFull)** for PR 10542 at commit [`ee29dd2`](https://github.com/apache/spark/commit/ee29dd230702413238abed77fc6a7691bd4ff6a5).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/48613/
Test FAILed.

holdenk commented:
Created a fix in https://github.com/apache/spark/pull/10564 , seems that the lag between the test being written and it getting merged had some bits change underneath it.

gatorsmile commented:
Thank you! @holdenk

holdenk commented:
Sure thing - thanks for pinging me when you noticed the issue :)

rxin committed:
Update MimaExcludes now Spark 1.6 is in Maven.

Author: Reynold Xin <rxin@databricks.com>

Closes #10561 from rxin/update-mima.

Cazen committed:
[SPARK-12537][SQL] Add option to accept quoting of all character backslash quoting mechanism

We can provides the option to choose JSON parser can be enabled to accept quoting of all character or not.

Author: Cazen <Cazen@korea.com>
Author: Cazen Lee <cazen.lee@samsung.com>
Author: Cazen Lee <Cazen@korea.com>
Author: cazen.lee <cazen.lee@samsung.com>

Closes #10497 from Cazen/master.

holdenk committed:
[SPARK-12611][SQL][PYSPARK][TESTS] Fix test_infer_schema_to_local

Previously (when the PR was first created) not specifying b= explicitly was fine (and treated as default null) - instead be explicit about b being None in the test.

Author: Holden Karau <holden@us.ibm.com>

Closes #10564 from holdenk/SPARK-12611-fix-test-infer-schema-local.

gatorsmile commented:
retest this please.

SparkQA commented:
**[Test build #48628 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/48628/consoleFull)** for PR 10542 at commit [`ee29dd2`](https://github.com/apache/spark/commit/ee29dd230702413238abed77fc6a7691bd4ff6a5).

gatorsmile committed:
code refactoring and code merge

gatorsmile committed:
Merge remote-tracking branch 'origin/outerJoinConversion' into outerJoinElimination

# Conflicts:
#	sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

gatorsmile commented:
Let me close it and then resubmit the PR.

SparkQA commented:
**[Test build #48632 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/48632/consoleFull)** for PR 10542 at commit [`63d5d62`](https://github.com/apache/spark/commit/63d5d62bcc719fc988510281d6ed0b4add4eaa3c).

SparkQA commented:
**[Test build #48628 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/48628/consoleFull)** for PR 10542 at commit [`ee29dd2`](https://github.com/apache/spark/commit/ee29dd230702413238abed77fc6a7691bd4ff6a5).
 * This patch **fails from timeout after a configured wait of \`250m\`**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/48628/
Test FAILed.

SparkQA commented:
**[Test build #48632 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/48632/consoleFull)** for PR 10542 at commit [`63d5d62`](https://github.com/apache/spark/commit/63d5d62bcc719fc988510281d6ed0b4add4eaa3c).
 * This patch **fails from timeout after a configured wait of \`250m\`**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/48632/
Test FAILed.
